### PR TITLE
Parse OAuth Authorization header when request omits client secret

### PR DESCRIPTION
**Pull request by hickford**
This fixes error "unauthorized_client: invalid client secret" when client includes secret in Authorization header rather than request body. OAuth spec permits both: https://www.rfc-editor.org/rfc/rfc6749#section-2.3.1

> Clients in possession of a client password MAY use the HTTP Basic authentication scheme ... Alternatively, the authorization server MAY support including the   client credentials in the request-body

Sanity validation that client id and client secret in request are consistent with Authorization header.

Improve error descriptions. Error codes remain the same.

* * *

I believe further improvements to OAuth remain necessary, in particular explicit client type #21299.

### CODE REVIEW

**Comment by hickford** [comment]
@wxiaoguang  This suffices for https://github.com/GitCredentialManager/git-credential-manager/pull/879

**Comment by wxiaoguang** [comment]
It also looks good to me. Could there be some details about how this PR fixes the problem? Then the details can be put in the commit message and maintainers in the future can also understand it.

And a small question about the ErrorDescription, will they be used by clients or should they be kept stable (never-changed)?

**Review by wxiaoguang — APPROVED** [review]

File: routers/web/auth/oauth.go [hunk 1]
```diff
@@ -588,7 +588,7 @@ func OIDCKeys(ctx *context.Context) {
 // AccessTokenOAuth manages all access token requests by the client
 func AccessTokenOAuth(ctx *context.Context) {
 	form := *web.GetForm(ctx).(*forms.AccessTokenForm)
-	if form.ClientID == "" {
+	if form.ClientID == "" || form.ClientSecret == "" {
```
Comment (wxiaoguang) [inline routers/web/auth/oauth.go:592]: Maybe some comments like:

```suggestion
	// if there is no ClientID or ClientSecret in the request body, fill these fields by the Authorization header and ensure the provided field matches the Authorization header
	if form.ClientID == "" || form.ClientSecret == "" {
```

**Comment by hickford** [comment]
Added details to  commit message and PR description.

The prescribed error codes remain the same. I change the human-readable description to "provide additional information" https://www.rfc-editor.org/rfc/rfc6749#section-5.2

**Review by delvh — APPROVED** [review]

**Review by KN4CK3R — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
I think it's worth to be backported to 1.17, how do you think?

**Comment by hickford** [comment]
Thanks. I think it would also be necessary to backport https://github.com/go-gitea/gitea/pull/21293 .

**Pull request merged**